### DISCUSSION TITLE
How to use `pifpaf`  and PostgreSQL with SQLAlchemy 2.0.18+?

### DISCUSSION THREAD

**Discussion by vytas7** [sqlalchemy/sqlalchemy#10076] [Usage Questions] [answered]
For running some tests we are using `pifpaf` to spin up a temporary Postgresql database and run `pytest` under it. If I'm not mistaken even @zzzeek mentioned he was using that somewhere too.

Maybe not entirely correct, but it used to work (including SQLAlchemy 2.0.17 and below) great for the lazy: just use the URL from the `PIFPAF_POSTGRESQL_URL` envvar verbatim:
```python
>>> import os, sqlalchemy; sqlalchemy.create_engine(os.environ['PIFPAF_POSTGRESQL_URL'])
Engine(postgresql://localhost/postgres?host=%2Ftmp%2Ftmpwfe2y0yu&port=9824)
```

However, it no longer works with SQLAlchemy 2.0.18:
```
<...>
sqlalchemy.exc.ArgumentError: Can't combine fixed host and multihost URL formats
```

**Comment by vytas7** [comment]
Aha is this potentially tracked as a bug and fixed in #10069?

**Reply by CaselIT** [reply]
That seems different.
Try removing the localhost from the url

**Reply by vytas7** [reply]
Hm you'd be right, I tried to install from Git `main` and I'm still getting the same problem.

Test case used (`pip install` the desired versions of `sqlalchemy`, `pifpaf` and `psycopg2`):
```
pifpaf run postgresql -- python -c "import os, sqlalchemy; print(sqlalchemy.__version__, sqlalchemy.create_engine(os.environ['PIFPAF_POSTGRESQL_URL']))"
```

Desired result:
```
2.0.17 Engine(postgresql://localhost/postgres?host=%2Ftmp%2Ftmp7_rimcfi&port=9824)
```

@CaselIT how should my URL look like then?
Maybe it would somehow possible to revert or to improve the code to afford this pattern again?

Thinking that it feels a bit strange not to support the URL format that the official `psql` command line client happily accepts:
```
$ psql "postgresql://localhost/postgres?host=/tmp/tmpjopzr284&port=9824" -c "show data_directory;"
  data_directory  
------------------
 /tmp/tmpjopzr284
(1 row)
```

**Reply by CaselIT** [reply]
That's unfortunate.

I think the current code expects `postgresql:///postgres?host=%2Ftmp%2Ftmp7_rimcfi&port=982`

@zzzeek this change seems very fragile. How about we parse things only and let the driver raise if it does not support what was passed?

**Reply by zzzeek** [reply]
im just looking at this now, what's the URL encoding about ?   i dont understand that?  we pass through host and it should be the same as it always was

**Reply by zzzeek** [reply]
why would the URL have both "localhost" and host=somehost in it?  which host do we use?  how is it not an error for us to discard one ?

> Thinking that it feels a bit strange not to support the URL format that the official psql command line client happily accepts:

we aren't psql, this is SQLAlchemy's `create_engine()` call which accepts a platform-agnostic URL format

**Reply by zzzeek** [reply]
postgresql literally silently ignores the first host:

```
[classic@zotac sqlalchemy:main]$ psql postgresql://scott:tiger@pg12/postgres
psql (15.1, server 12.12)
Type "help" for help.

postgres=> ^C
postgres=> 
\q
[classic@zotac sqlalchemy:main]$ psql postgresql://scott:tiger@pg12/postgres?host=pg13
psql (15.1, server 13.8)
Type "help" for help.

```

wow I HATE that, what's wrong with them?

**Reply by zzzeek** [reply]
yeah.  sure, I will take that whole thing out.  silently ignore the host, OK

**Reply by vytas7** [reply]
I agree that what `pifpaf` is providing in that `$PIFPAF_POSTGRESQL_URL` does sound confusing, and it is interesting that `psql` happily accepts that.

I'm more looking after an easy/lazy way to use this with `pifpaf` again though :)

**Reply by zzzeek** [reply, marked as answer]
just block SQLAlchemy 2.0.18 from your requirements and use 2.0.17 and/or 2.0.19.

**Reply by vytas7** [reply]
So in a pinch something along the above line works for me 
```
$ pifpaf run postgresql -- python -c "import os, sqlalchemy; print(sqlalchemy.__version__, sqlalchemy.create_engine(os.environ['PIFPAF_POSTGRESQL_URL'].replace('localhost', '')))"
2.0.19 Engine(postgresql:///postgres?host=%2Ftmp%2Ftmpxlvm9gyr&port=9824)
```
(But yeah maybe this was a bit abrupt change as for a micro version.)

**Reply by CaselIT** [reply]
> postgresql literally silently ignores the first host:

I think the reasoning is that the query parameter overrides the host in the url.
Still it seems unfortunate
